src/Features/CSharp/Portable/CodeRefactorings/ExtractClass/CSharpExtractClassCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,44 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.CodeRefactorings.PullMemberUp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.ExtractClass;
+using Microsoft.CodeAnalysis.Host.Mef;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.ExtractClass
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(PredefinedCodeRefactoringProviderNames.ExtractClass)), Shared]
+    internal class CSharpExtractClassCodeRefactoringProvider : AbstractExtractClassRefactoringProvider
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CSharpExtractClassCodeRefactoringProvider()
+            : base()
+        {
+        }
+
+        /// <summary>
+        /// Test purpose only.
+        /// </summary>
+        [SuppressMessage("RoslynDiagnosticsReliability", "RS0034:Exported parts should have [ImportingConstructor]", Justification = "Used incorrectly by tests")]
+        internal CSharpExtractClassCodeRefactoringProvider(IExtractClassOptionsService optionsService)
+            : base(optionsService)
+        {
+        }
+
+        protected override async Task<SyntaxNode?> GetSelectedClassDeclarationAsync(CodeRefactoringContext context)
+            => (await context.GetRelevantNodesAsync<ClassDeclarationSyntax>().ConfigureAwait(false)).FirstOrDefault();
+
+        protected override Task<SyntaxNode?> GetSelectedNodeAsync(CodeRefactoringContext context) => CSharpPullMemberUpCodeRefactoringProvider.GetSelectedDeclarationOrVariableAsync(context);
```

Comment:
```suggestion
        protected override Task<SyntaxNode?> GetSelectedNodeAsync(CodeRefactoringContext context)
            => CSharpPullMemberUpCodeRefactoringProvider.GetSelectedDeclarationOrVariableAsync(context);
```

Reply:
i don't love the features calling into each otehr.  i would prefer we extract our those helpers to a new common location.  like "UseNodeSelectionHelpers".   or something

Reply:
I missed this last pass, will include moving out as a change next iteration.

Reply:
Added `NodeSelectionHelpers`